src/app/api/v1/location-autocomplete/route.ts

```diff
@@ -0,0 +1,114 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { NextResponse, NextRequest } from "next/server";
+import uFuzzy from "@leeoniya/ufuzzy";
+
+import { IRelevantLocation } from "../../../../scripts/createLocationAutocompleteData/types.d";
+import locationData from "../../../../../locationAutocompleteData.json";
+
+export type TMatchingLocations = Array<IRelevantLocation> | [];
+
+export interface ISearchLocationParams {
+  searchQuery: string;
+  config: {
+    minQueryLength: number;
+    maxResults: number;
+  };
+}
+
+export interface ISearchLocationResults {
+  searchQuery: string;
+  results: TMatchingLocations;
+}
+
+async function getLocationsByQuery(searchQuery: string) {
+  const locationNames = locationData.data.map((location) => {
+    const { name, stateCode, countryCode, alternateNames } = location;
+    const alternateNamesJoined = alternateNames
+      ? location.alternateNames.join(" ")
+      : "";
+
+    return `${name} ${stateCode} ${countryCode} ${alternateNamesJoined}`;
+  });
+
+  // For search options see: https://github.com/leeoniya/uFuzzy#options
+  const fuzzySearch = new uFuzzy({
+    intraMode: 1,
+    intraIns: 1,
+    interIns: 3,
+    intraSub: 1,
+    intraTrn: 1,
+    intraDel: 1,
+    // matches lowercase letters, whitespace or apostrophe
+    intraChars: "[a-z\\s\\d']",
+    interLft: 2,
+    interRgt: 0,
+  });
+
+  const locationIndexes = fuzzySearch.filter(locationNames, searchQuery);
+  if (!locationIndexes) {
+    return [];
+  }
+
+  const info = fuzzySearch.info(locationIndexes, locationNames, searchQuery);
+  const order = fuzzySearch.sort(info, locationNames, searchQuery);
+  const results = locationIndexes.map(
+    (locationIndex) => locationData.data[locationIndex]
+  );
+
+  const resultsOrdered = order.map((orderIndex) => results[orderIndex]);
+  // Split and only sort the first x percentage of results by population.
```

Comment:
Could you expand on this comment why we're doing this? My interpretation is that this takes the top 75% most relevant results, but then ranks the ones with the largest population first - wouldn't that potentially lead to the closest match only being 75% down the list?

Reply:
That’s right. The idea is to improve the results and take the population into account to improve the “scores” for matches, but not including very weak matches. This especially improves the suggestions for short search queries that can have a lot of matches.

Expanded on the comment: [339d004](https://github.com/mozilla/blurts-server/pull/3186/commits/339d0043408d90becb51514969ea1113a17fb916).